GoogleForCreators/web-stories-wp
# [TASK] Foundation for interpreting story animations schema

## Summary
- Adds `StoryAnimation.Provider`
  - Interprets animations into generators for a given ID, rn used with WAAPI, but compatible with AMP
  - Adds a `playWAAPIAnimations()` method to control WAAPI animations
  - Adds an `onWAAPIFinish` prop that will fire everytime all hoisted animations complete (can use to increment page number)
- Adds `StoryAnimation.WAAPIWrapper`
  - Composes all animations for a target with later animations being children of earlier anims
  - Creates WAAPI animation part based off of type.
- Adds WAAPI animation part generator which given a type and that types args, creates & hoists animation as well as creating relevant markup for animation.
- Creates notion of `Animation Parts` which ties the `keyframe` & `timings` to markup given some args. right now only did this for `ANIMATION_TYPES.BOUNCE` but should be simple to port over old animation configs to this new pattern.

**example use:**
```
<StoryAnimation.Provider animations={[
  ...,
  { target: "some_uuid", type: <animation_type>, ...args }
]}>
...
  <div data-element-id="some_uuid" class="wrapper">
      <StoryAnimation.WAAPIWrapper target="some_uuid">
        <div />
      </StoryAnimation.WAAPIWrapper>
  </div>
...
</StoryAnimation.Provider>
```

## Relevant Technical Choices
**Web Animations API - Grouping**
Basically had to add some notion of grouping on top of the web animations api. Accomplished this 
by just hoisting all `Animation` objects up into the provider. We can then create any group operations we'd like on the hoisted animations. rn only `play` & `onfinish` supported.

**Animation Parts**
Animation schema is basically looks like:
`{ type: 'ANIM_TYPE', ...args }` which we transform into `ANIM_TYPE(args) => ReactWrapperComponent`. This should be a pretty extensible in the future for applying to amp-animations or any other animation package. Possibly maleable enough to add a `ANIMATION_TYPE.SEQUENCE` or anything else in the future because we just pass relevant `args` to a function that creates a react component.


## To-do
- Port over rest of animation configs to animation parts and add containers
- Create `<StoryAnimation.AMPWrapper />` which follows the animation parts pattern
- Create `AMP(type, args)` animation part generator (should be able to leverage @mariano-formidable amp-animation animator work here)
- Create amp-animation methods in provider (should mostly be passing around animation id names which should be nice)

## How to test
**Unit**
 Run unit tests & make sure they pass.

**QA**
Open up storybook and go to `Dashboard > Components > StoryAnimation > default` this should show a square that has nested `BOUNCE` animations applied to it and alternates between red & green color each time all the animations finish


---
